README.md
# GBAWAVE
Most of this source code comes from [Pin Eight](https://pineight.com/gba/gsm/), my changes consist of adding images for each track.

Check out the full project description on [Tindie!](https://www.tindie.com/products/gbawave/gbawave-audio-custom-mixtapes-on-gameboy-advance/)


CHANGES.txt
GSM Player 2004-08-23:
  hud: switch from 6x8 to 8x16 pixel fonts; now much less ugly
  player: switch to devkitARM R8
  win32: improved Go.bat to take advantage of Win2K's better FOR command;
         Windows 98/ME users should use Go-Win9x.bat

GSM Player 2004-06-27:
  codec: unroll some loops; now CPU use is about 50 percent
  docs: clarify the result of the conversion
  player: first crack at fixing exit-to-menu

GSM Player 2004-03-06:
  initial release of ported codec and player

Future:
  cubic interpolation and noise-shaped dithering of audio
  long file name extensions to player
  even less ugly interface

hud.c
#include <stdlib.h>
#include "pin8gba.h"
#include "gbfs.h"

extern const char _8x16_fnt[];
extern const unsigned int _8x16_fnt_len;

extern u32 fracumul(u32, u32) __attribute__((long_call));
extern s32 dv(s32, s32) __attribute__((long_call));

/* upcvt_4bit() ************************
   Converts a 1-bit font to GBA 4-bit format.
*/
void upcvt_4bit(void *dst, const u8 *src, size_t len)
{
  u32 *out = dst;

  for(; len > 0; len--)
  {
    u32 dst_bits = 0;
    u32 src_bits = *src++;
    u32 x;

    for(x = 0; x < 8; x++)
    {
      dst_bits <<= 4;
      dst_bits |= src_bits & 1;
      src_bits >>= 1;
    }
    *out++ = dst_bits;
  }
}

static void hud_wline(unsigned int y, const char *s)
{
  unsigned short *dst = MAP[31][y * 2] + 1;
  unsigned int wid_left;

  for(wid_left = 28; wid_left > 0 && *s; wid_left--)
  {
    unsigned char c0 = *s++;

    dst[0] = c0 << 1;
    dst[32] = (c0 << 1) | 1;
    ++dst;
  }
  for(; wid_left > 0; wid_left--, dst++)
  {
    dst[0] = 0x0040;
    dst[32] = 0x0041;
  }
}

static void hud_cls(void)
{
  unsigned int i;

  for(i = 0; i < 32*20; i++)
    MAP[31][0][i] = 0x0040;

  /*hud_wline(1, "GSM Player for GBA");
  hud_wline(2, "by Damian Yerrick");*/
  //I'm sorry Damian, I love your code, but I gotta do it for the aesthetic!
}

void hud_init(void)
{
  PALRAM[0] = RGB(0,0,0);
  PALRAM[1] = RGB(31,31,31);
  LCDMODE = 0;
  upcvt_4bit(VRAM, _8x16_fnt, _8x16_fnt_len);

  BGCTRL[2] = BGCTRL_NAME(31) | BGCTRL_PAT(0);

  hud_cls();
  /*hud_wline(3, "Copyright 2004");

  hud_wline(5, "For more info see");
  hud_wline(6, "TOAST-COPYRIGHT.txt");

  while(LCD_Y < 160) {}*/
  LCDMODE = 0 | LCDMODE_BG2;
}



/* base 10, 10, 6, 10 conversion */
static unsigned int hud_bcd[] =
{
  600, 60, 10, 1  
};


#undef BCD_LOOP
#define BCD_LOOP(b) if(src >= fac << b) { src -= fac << b; c += 1 << b; }

static void decimal_time(char *dst, unsigned int src)
{
  unsigned int i;

  for(i = 0; i < 4; i++)
    {
      unsigned int fac = hud_bcd[i];
      char c = '0';

      BCD_LOOP(3);
      BCD_LOOP(2);
      BCD_LOOP(1);
      BCD_LOOP(0);
      *dst++ = c;
    }
}

struct HUD_CLOCK
{
  unsigned int cycles;
  unsigned char trackno[2];
  unsigned char clock[4];
} hud_clock;

static const char clockmax[4] = {9, 9, 5, 9};

void hud_frame(int locked, unsigned int t)
{
  char line[16];
  char time_bcd[4];

  /* a fractional value for Seconds Per Byte
     1/33 frame/byte * 160 sample/frame * 924 cpu/sample / 2^24 sec/cpu
     * 2^32 fracunits = 1146880 sec/byte fracunits
   */

  t = fracumul(t, 1146880);
  if(t > 5999)
    t = 5999;
  decimal_time(time_bcd, t);

  line[0] = (locked & JOY_SELECT) ? 12 : ' ';
  line[1] = (locked & JOY_START) ? 16 : ' ';
  line[2] = ' ';
  line[3] = hud_clock.trackno[0] + '0';
  line[4] = hud_clock.trackno[1] + '0';
  line[5] = ' ';
  line[6] = ' ';
  line[7] = time_bcd[0];
  line[8] = time_bcd[1];
  line[9] = ':';
  line[10] = time_bcd[2];
  line[11] = time_bcd[3];
  line[12] = '\0';
  hud_wline(9, line);
}


/*void hud_new_song(const char *name, unsigned int trackno)
{
  int upper;

  hud_cls();
  hud_wline(4, "Now playing");
  hud_wline(5, name);
  hud_clock.cycles = 0;

  for(upper = 0; upper < 4; upper++)
    hud_clock.clock[0] = 0;
  upper = dv(trackno, 10);
  hud_clock.trackno[1] = trackno - upper * 10;
  trackno = upper;
  upper = dv(trackno, 10);
  hud_clock.trackno[0] = trackno - upper * 10;
}*/

void hud_new_song(const char *name, GBFS_FILE *fs){
	char imgName[strlen(name)+6];
	strcpy(imgName, "img");
	strcat(imgName, name);
	//while(LCD_Y >= 160);
	//while(LCD_Y < 160);
	gbfs_copy_obj(0x6000000, fs, imgName);
}

void bmp16_rect(int left, int top, int right, int bottom, u32 clr,
    void *dstBase, u32 dstPitch)
{
    int ix, iy;

    u32 width= right-left, height= bottom-top;
    u16 *dst= (u16*)(dstBase+top*dstPitch + left*2);
    dstPitch /= 2;

    // --- Draw ---
    for(iy=0; iy<height; iy++)
        for(ix=0; ix<width; ix++)
            dst[iy*dstPitch + ix]= clr;
}

isr.c
/*

Tetanus On Drugs for GBA
isr.c : interrupt service routine
Copyright (C) 2002  Damian Yerrick

This program is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License
as published by the Free Software Foundation; either version 2
of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to 
  Free Software Foundation, Inc., 59 Temple Place - Suite 330,
  Boston, MA  02111-1307, USA.
GNU licenses can be viewed online at http://www.gnu.org/copyleft/

Visit http://www.pineight.com/ for more information.

*/

volatile int want_reset = 0;

#include "pin8gba.h"

#define INTACK          (*(volatile u16 *)0x04000202)
#define BIOS_INTACK     (*(volatile u16 *)0x03007ff8)
#define INTMASK_VBL     0x0001

#define INTENABLE       (*(volatile u16 *)0x04000208)


void dsound_vblank(void);

void isr(void)
{
  unsigned int interrupts = INTACK;

#if 0
  if(interrupts & INTMASK_VBL)
  {
    dsound_vblank();
  }
#endif
  BIOS_INTACK |= interrupts;
  INTACK = interrupts;
  INTENABLE = 1;
}


config.h
/*
 * Copyright 1992 by Jutta Degener and Carsten Bormann, Technische
 * Universitaet Berlin.  See the accompanying file "COPYRIGHT" for
 * details.  THERE IS ABSOLUTELY NO WARRANTY FOR THIS SOFTWARE.
 */

/*$Header: /tmp_amd/presto/export/kbs/jutta/src/gsm/RCS/config.h,v 1.5 1996/07/02 11:26:20 jutta Exp $*/

#ifndef	CONFIG_H
#define	CONFIG_H

/*efine	SIGHANDLER_T	int 		/* signal handlers are void	*/
//efine HAS_SYSV_SIGNAL	1		/* sigs not blocked/reset?	*/

#define	HAS_STDLIB_H	1		/* /usr/include/stdlib.h	*/
//efine	HAS_LIMITS_H	1		/* /usr/include/limits.h	*/
#define	HAS_FCNTL_H	1		/* /usr/include/fcntl.h		*/
//efine	HAS_ERRNO_DECL	1		/* errno.h declares errno	*/

#define	HAS_FSTAT 	1		/* fstat syscall		*/
#define	HAS_FCHMOD 	1		/* fchmod syscall		*/
#define	HAS_CHMOD 	1		/* chmod syscall		*/
#define	HAS_FCHOWN 	1		/* fchown syscall		*/
#define	HAS_CHOWN 	1		/* chown syscall		*/
//efine	HAS__FSETMODE 	1		/* _fsetmode -- set file mode	*/

#define	HAS_STRING_H 	1		/* /usr/include/string.h 	*/
//efine	HAS_STRINGS_H	1		/* /usr/include/strings.h 	*/

#define	HAS_UNISTD_H	1		/* /usr/include/unistd.h	*/
#define	HAS_UTIME	1		/* POSIX utime(path, times)	*/
//efine	HAS_UTIMES	1		/* use utimes()	syscall instead	*/
#define	HAS_UTIME_H	1		/* UTIME header file		*/
//efine	HAS_UTIMBUF	1		/* struct utimbuf		*/
//efine	HAS_UTIMEUSEC   1		/* microseconds in utimbuf?	*/

#define CODE_IN_IWRAM __attribute__ ((section (".iwram"), long_call))

#endif	/* CONFIG_H */


tools/catbin.c
/*
catbin.c
Concatenate two or more binary files into one
By Damian Yerrick


Copyright 2001 Damian Yerrick

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
"Software"), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN
AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF
OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
IN THE SOFTWARE.

*/


/* Rationale

What is catbin?

  catbin is a program to take multiple binary files and produce the
  concatenation of those files.

Why not just use DOS copy?

  MinGW make doesn't seem to be able to call copy /b properly.

Why not just use cat?

  Neither MinGW nor the minimal Cygwin distribution included with
  devkit advance includes a port of GNU fileutils to win32.

DJGPP has fileutils.  Why not use DJGPP's cat?

  DJGPP and Windows 2000 have issues with each other.

Why not require the full Cygwin distribution?

  I prefer MinGW to Cygwin because MinGW is a smaller download.  I
  don't expect all users of my software to have upwards of $200,000
  to move their families to an area where broadband Internet access
  is affordable.

*/

#include <stdio.h>
#include <stdlib.h>

const char syntax_help[] =
"catbin 0.1 by Damian Yerrick: concatenates binary files\n"
"usage: catbin INFILE [INFILE...] OUTFILE\n";

int main(int argc, char **argv)
{
  char buf[1024];
  size_t n_got;
  int arg;
  FILE *infile, *outfile;

  if(argc < 3)
    {
      fputs(syntax_help, stderr);
      return 1;
    }

  outfile = fopen(argv[argc - 1], "wb");
  if(!outfile)
    {
      fputs("catbin could not open output file ", stderr);
      perror(argv[argc - 1]);
      return 1;
    }

  for(arg = 1; arg < argc - 1; arg++)
    {
      infile = fopen(argv[arg], "rb");
      if(!infile)
	{
	  fclose(outfile);
	  fputs("catbin could not open input file ", stderr);
	  perror(argv[arg]);
	  return 1;
	}
      while((n_got = fread(buf, sizeof(char), sizeof(buf), infile)) > 0)
	{
	  size_t n_written;

	  n_written = fwrite(buf, sizeof(char), n_got, outfile);
	  if(n_written < n_got)
	    {
	      fclose(infile);
	      fclose(outfile);
	      fprintf(stderr, "%lu bytes read but %lu bytes written\n",
		      (unsigned long)n_got, (unsigned long)n_written);
	      fputs("catbin could not write to input file ", stderr);
	      perror(argv[argc - 1]);
	      return 1;
	    }
	}
    }

  fclose(outfile);
  return 0;
}


tools/padbin.c
/* padbin.c
   pad a binary to a given size boundary

Copyright (C) 2002  Damian Yerrick

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
"Software"), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN
AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF
OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
IN THE SOFTWARE.

*/


#include <stdio.h>
#include <stdlib.h>

char zeroz[1024];

int main(int argc, char **argv)
{
  FILE *fp;
  unsigned long overage;
  unsigned long factor;

  /* clear to 0xff for faster flash writing */
  for(factor = 0; factor < 1024; factor++)
    zeroz[factor] = 0xff;

  if(argc != 3)
  {
    fputs("pads a binary file to an integer multiple of a given number of bytes\n"
          "syntax: padbin FACTOR FILE\n"
          "FACTOR can be decimal (e.g. 256), octal (e.g. 0400), or hex (e.g. 0x100)\n", stderr);
    return 1;
  }

  factor = strtoul(argv[1], NULL, 0);
  if(factor < 2)
  {
    fputs("error: FACTOR must be greater than or equal to 2\n", stderr);
    return 1;
  }

  fp = fopen(argv[2], "rb+");
  if(!fp)
  {
    fputs("could not open", stderr);
    perror(argv[2]);
    return 1;
  }

  /* find the amount the file has over the limit */
  fseek(fp, 0, SEEK_END);
  overage = ftell(fp) % factor;
  if(overage != 0)
  {
    unsigned long extension = factor - overage;

    while(extension >= 1024)
    {
      fwrite(zeroz, 1, 1024, fp);
      extension -= 1024;
    }
    if(extension >= 0)
    {
      fwrite(zeroz, 1, extension, fp);
    }
  }
  fclose(fp);
  return 0;
}


gsmplay.c
/* gsmplay.c
   Plays GSM audio through Game Boy Advance hardware.
*/

/*
 * Copyright 2004 by Damian Yerrick.
 * See the accompanying file "TOAST-COPYRIGHT.txt" for details.
 * THERE IS ABSOLUTELY NO WARRANTY FOR THIS SOFTWARE.
 */

#include <stdlib.h>
#include <string.h>
#include "pin8gba.h"
#include "gsm.h"
#include "private.h" /* for sizeof(struct gsm_state) */

#include "gbfs.h"

void isr(void);

#if 0
#define PROFILE_WAIT_Y(y) \
	do                    \
	{                     \
	} while (LCD_Y != (y))
#define PROFILE_COLOR(r, g, b) (PALRAM[0] = RGB((r), (g), (b)))
#else
#define PROFILE_WAIT_Y(y) ((void)0)
#define PROFILE_COLOR(r, g, b) ((void)0)
#endif

static void dsound_switch_buffers(const void *src)
{
	DMA[1].control = 0;

	/* no-op to let DMA registers catch up */
	asm volatile("eor r0, r0; eor r0, r0" ::
					 : "r0");

	DMA[1].src = src;
	DMA[1].dst = (void *)0x040000a0; /* write to FIFO A address */
	DMA[1].count = 1;
	DMA[1].control = DMA_DSTUNCH | DMA_SRCINC | DMA_REPEAT | DMA_U32 |
					 DMA_SPECIAL | DMA_ENABLE;
}

void init_sound(void)
{
	TIMER[0].control = 0;
	//turn on sound circuit
	SETSNDRES(1);
	SNDSTAT = SNDSTAT_ENABLE;
	DSOUNDCTRL = 0x0b0e;
	TIMER[0].count = 0x10000 - (924 / 2);
	TIMER[0].control = TIMER_16MHZ | TIMER_ENABLE;
}

/* gsm_init() **************
   This is to gsm_create() as placement new is to new.
*/
void gsm_init(gsm r)
{
	memset((char *)r, 0, sizeof(*r));
	r->nrp = 40;
}

void wait4vbl(void)
{
	asm volatile("mov r2, #0; swi 0x05" ::
					 : "r0", "r1", "r2", "r3");
}

struct gsm_state decoder;
const GBFS_FILE *fs;
const char *src;
unsigned int src_len;

signed short out_samples[160];
signed char double_buffers[2][608] __attribute__((aligned(4)));

#if 0

static void dsound_silence(void)
{
  DMA[1].control = 0;
}

void pre_decode_run(void)
{
  const char *src_pos;
  const char *src_end;
  char *dst_pos = EWRAM;

  src = gbfs_get_obj(fs, "butterfly.pcm.gsm", &src_len);
  if(!src)
    {
      LCDMODE = 0;
      PALRAM[0] = RGB(31, 23, 0);
      while(1) {}
    }

  src_pos = src;
  src_end = src_pos + src_len;

  gsm_init(&decoder);

  while(src_pos < src_end)
    {
      unsigned int i;

      while(LCD_Y != 0 && LCD_Y != 76 && LCD_Y != 152) {}

      PALRAM[0] = RGB(0, 0, 31);
      if(gsm_decode(&decoder, src_pos, out_samples))
	{
	  LCDMODE = 0;
	  PALRAM[0] = RGB(31, 31, 0);
	  while(1) {}
	}
      src_pos += sizeof(gsm_frame);
      PALRAM[0] = RGB(0, 0, 15);
      for(i = 0; i < 160; i++)
	dst_pos[i] = out_samples[i] >> 8;
      dst_pos += 160;
      PALRAM[0] = RGB(0, 0, 0);
    }
  PALRAM[0] = RGB(0, 31, 0);

  dsound_switch_buffers(EWRAM);

}
#endif

#define CMD_START_SONG 0x0400

//void reset_gba(void) __attribute__((long_call));
void hud_init(void);
void hud_new_song(const char *name, GBFS_FILE *fs);
void hud_frame(int locked, unsigned int t);
void bmp16_rect(int left, int top, int right, int bottom, u32 clr, void *dstBase, u32 dstPitch);

void streaming_run(void)
{
	const char *src_pos = NULL;
	const char *src_end = NULL;
	unsigned int decode_pos = 160, cur_buffer = 0;
	unsigned short last_joy = 0x3ff;
	unsigned int cur_song = (unsigned int)(-1);
	int last_sample = 0;
	int locked = 0;

	while (1)
	{
		unsigned short j = (JOY & 0x3ff) ^ 0x3ff;
		unsigned short cmd = j & (~last_joy | JOY_R | JOY_L);
		signed char *dst_pos = double_buffers[cur_buffer];

		last_joy = j;

		/*      if((j & (JOY_A | JOY_B | JOY_SELECT | JOY_START))
         == (JOY_A | JOY_B | JOY_SELECT | JOY_START))
        reset_gba(); */

		if (cmd & JOY_SELECT)
			locked ^= JOY_SELECT;

		if (locked & JOY_SELECT)
			cmd = 0;

		if (cmd & JOY_START)
			locked ^= JOY_START;

		if (cmd & JOY_L)
		{
			src_pos -= 33 * 50;
			if (src_pos < src)
				cmd |= JOY_LEFT;
		}

		if (cmd & JOY_R)
		{
			src_pos += 33 * 50;
		}

		if (src_pos >= src_end)
			cmd |= JOY_RIGHT;

		if (cmd & JOY_RIGHT)
		{
			cur_song++;
			if (cur_song >= (gbfs_count_objs(fs)/2))
				cur_song = 0;
			cmd |= CMD_START_SONG;
		}

		if (cmd & JOY_LEFT)
		{
			if (cur_song == 0)
				cur_song = (gbfs_count_objs(fs)/2) - 1;
			else
				cur_song--;
			cmd |= CMD_START_SONG;
		}

		if (cmd & CMD_START_SONG)
		{
			char name[25];
			gsm_init(&decoder);
			src = gbfs_get_nth_obj(fs, cur_song, name, &src_len);
			//hud_new_song(name, cur_song + 1);
			hud_new_song(name, fs);
			if (cmd & JOY_L)
				src_pos = src + src_len - 33 * 60;
			else
				src_pos = src;
			src_end = src + src_len;
		}

		PROFILE_WAIT_Y(0);

		// PALRAM[0] = RGB(22, 0, 0);

		if (locked & JOY_START) /* if paused */
			for (j = 304 / 2; j > 0; j--)
			{
				*dst_pos++ = last_sample >> 8;
				*dst_pos++ = last_sample >> 8;
				*dst_pos++ = last_sample >> 8;
				*dst_pos++ = last_sample >> 8;
			}
		else
			for (j = 304 / 4; j > 0; j--)
			{
				int cur_sample;
				if (decode_pos >= 160)
				{
					if (src_pos < src_end)
						gsm_decode(&decoder, src_pos, out_samples);
					src_pos += sizeof(gsm_frame);
					decode_pos = 0;
				}

				/* 2:1 linear interpolation */
				cur_sample = out_samples[decode_pos++];
				*dst_pos++ = (last_sample + cur_sample) >> 9;
				*dst_pos++ = cur_sample >> 8;
				last_sample = cur_sample;

				cur_sample = out_samples[decode_pos++];
				*dst_pos++ = (last_sample + cur_sample) >> 9;
				*dst_pos++ = cur_sample >> 8;
				last_sample = cur_sample;

				cur_sample = out_samples[decode_pos++];
				*dst_pos++ = (last_sample + cur_sample) >> 9;
				*dst_pos++ = cur_sample >> 8;
				last_sample = cur_sample;

				cur_sample = out_samples[decode_pos++];
				*dst_pos++ = (last_sample + cur_sample) >> 9;
				*dst_pos++ = cur_sample >> 8;
				last_sample = cur_sample;
			}

		PROFILE_COLOR(27, 27, 27);
		wait4vbl();
		dsound_switch_buffers(double_buffers[cur_buffer]);
		PROFILE_COLOR(27, 31, 27);

		bmp16_rect(4,149, 4+(232*((src_pos-src)/((double)src_len))), 154, 0, 0x6000000, 480);
		//hud_frame(locked, src_pos - src); TODO: Add Progress bar here?
		cur_buffer = !cur_buffer;
	}
}

void copr_wait(void)
{
	unsigned int vbls = 360;

	do
	{
		wait4vbl();
	} while (--vbls);
}

int main(void)
{
	/* enable interrupts */
	SET_MASTER_ISR(isr);
	LCDSTAT = LCDSTAT_VBLIRQ;			 /* one plug to the display */
	INTMASK = INT_VBLANK | INT_TIMER(1); /* the other to the isr */
	INTENABLE = 1;						 /* and flip the switch */
	fs = find_first_gbfs_file(find_first_gbfs_file);
	if (!fs)
	{
		LCDMODE = 0;
		PALRAM[0] = RGB(31, 0, 0);
		while (1){}
	}
	LCDMODE = 0x0400 | 0x0003; //Set Screen to mode three
	//gbfs_copy_obj(0x6000000, fs, "test1");
	//hud_init();
	//copr_wait();
	init_sound();

	streaming_run();
	while(1){}
}
